### ISSUE TITLE
(design) Change the order of block processing

### ISSUE THREAD

**Issue by tlienart** [tlienart/Franklin.jl#703]
This, for now, is just a brainstorm of what could, potentially, be a significant refactoring of a key part of Franklin which may, ultimately lead to wiping out the annoying issues of  stray `<p>` cleanly. 

## Current processing

Basically the  main passes are:

* tokenization + forming of blocks (from-->to) 
* extracting a markdown text (A) where all special blocks have been removed and replaced by a filler
* † go over (A) gradually, weaving "base text" using Julia's Markdown to convert markdown to html and Franklin's conversion of  special blocks to html
* a pass over the HTML to resolve hfuns etc. 

The step marked with † is the one that is prone to generating html with problems. Here's a naive illustration of the basic steps:

**input**

````
A
@@name B @@
C
````

**intermediate md**

```
A
#FILLER#
B
```

**parts** 

* `A` --> `<p>A</p>`
* `#FILLER#` --> resolution of `@@name B @@` --> `<div class="name">B</div>`
* `B` --> `<p>B</p>`

**weaved html**

```
<p>A</p>
<div class="name">B</div>
<p>B</p>
```

In this case it works fine; but as it gets more complicated; it can become unclear whether the markdown bits should close `</p>` or not. This is particularly true in the case of nesting. 

## Alternative route

An alternative route would be to do  the following:

* resolve the blocks first
* write intermediate **markdown** with raw html injected in  it
* run a Markdown to html pass

So in the previous example this would be

**intermediate markdown**

````
A
```{=html}
<div class="name">B</div>
```
C
````

**conversion pass** (e.g. with CommonMark's `ast = parser(s); html(ast)`)

```
<p>A</p>
<div class="name">B</div>
<p>C</p>
```

So here the result is identical. What changes is that we let a sane Markdown -> HTML process happen which takes care of the `p` etc. 

The other note is that this is the path to  using CommonMark (#498) and thereby get access to a bunch of goodies like proper handling of lists and tables. 

## Things to check

Basically we need to check that html raw blocks can be injected anywhere in CommonMark markdown without problems in particular:

* in lists
* in tables

also figuring out the difference between inline html and block html for CommonMark.

## Conclusion

This might just work but it'll take some time to put in place + test to ensure it doesn't break stuff. I'll probably start working on it slowly on a separate branch and report progress (or major blockers) here over time... I'm thinking that if I make good enough progress with this, I'm willing to push back on a 1.0 release so that this gets in as it'd be much cleaner (would generate valid html most of the times).

**Comment by tlienart** [comment]
Some additional notes

* would need to disable processing of tabs https://github.com/MichaelHatherly/CommonMark.jl/issues/1#issuecomment-631488261
* there's a distinction between inline html and block html in CommonMark.jl which will likely be annoying for instance it seems that in a list item element, online inline html can be read but that means that at injection time we need to figure out that we're in a li context.
